src/sync/chase_lev.rs

```diff
@@ -158,69 +155,93 @@ impl<T> Deque<T> {
     unsafe fn push(&self, data: T) {
         let guard = epoch::pin();
 
-        let mut b = self.bottom.load(SeqCst);
-        let t = self.top.load(SeqCst);
-        let mut a = self.array.load(SeqCst, &guard).unwrap();
-        let size = b - t;
+        let mut b = self.bottom.load(Relaxed);
+        let t = self.top.load(Acquire);
+        let mut a = self.array.load(Relaxed, &guard).unwrap();
+        
+        let size = b.wrapping_sub(t);
         if size >= (a.size() as isize) - 1 {
             // You won't find this code in the chase-lev deque paper. This is
             // alluded to in a small footnote, however. We always free a buffer
             // when growing in order to prevent leaks.
             a = self.swap_buffer(b, a, a.resize(b, t, 1), &guard);
-            b = self.bottom.load(SeqCst);
+            
+            // reload the bottom counter, since swap_buffer modifies it.
+            b = self.bottom.load(Relaxed);
         }
         a.put(b, data);
-        self.bottom.store(b + 1, SeqCst);
+        fence(Release);
+        self.bottom.store(b.wrapping_add(1), Relaxed);
     }
 
     unsafe fn try_pop(&self) -> Option<T> {
+        // try to perform an early return.
+        {
+            let t = self.top.load(Relaxed);
+            let b = self.bottom.load(Relaxed);
+            let len = b.wrapping_sub(t);
+            if len <= 0 {
+                return None;
+            }
+        }
+        
         let guard = epoch::pin();
 
-        let b = self.bottom.load(SeqCst);
-        let a = self.array.load(SeqCst, &guard).unwrap();
-        let b = b - 1;
-        self.bottom.store(b, SeqCst);
-        let t = self.top.load(SeqCst);
-        let size = b - t;
-        if size < 0 {
-            self.bottom.store(t, SeqCst);
-            return None
-        }
-        let data = a.get(b);
-        if size > 0 {
-            self.maybe_shrink(b, t, &guard);
-            return Some(data)
-        }
-        if self.top.compare_and_swap(t, t + 1, SeqCst) == t {
-            self.bottom.store(t + 1, SeqCst);
-            return Some(data)
+        let b = self.bottom.load(Relaxed).wrapping_sub(1);
+        let a = self.array.load(Relaxed, &guard).unwrap();
+        self.bottom.store(b, Relaxed);
+        fence(SeqCst); // the store to bottom must occur before loading top.
+        let t = self.top.load(Relaxed);
+        
+        let size = b.wrapping_sub(t);
+        if size >= 0 {
```

Comment:
Also: the two are equivalent in this case, since we first decrement b and then compare it to t. If they are equal we can store b+1 or t+1 at the end, it makes no difference.